R/score.R

```diff
@@ -217,14 +217,15 @@ score.forecast_quantile <- function(forecast, metrics = metrics_quantile(), ...)
 #' @return A data table with the forecasts and the calculated metrics.
 #' @keywords internal
 apply_metrics <- function(forecast, metrics, ...) {
-  expr <- expression(
-    forecast[, (metric_name) := do.call(run_safely, list(..., fun = fun))]
-  )
-  lapply(seq_along(metrics), function(i, forecast, ...) {
-    metric_name <- names(metrics[i]) # nolint
-    fun <- metrics[[i]] # nolint
-    eval(expr)
-  }, forecast, ...)
+  lapply(names(metrics), function(metric_name) {
+    result <- do.call(
+      run_safely,
+      list(..., fun = metrics[[metric_name]], metric_name = metric_name)
+    )
+    if (!is.null(result)) {
+      forecast[, (metric_name) := result]
+    }
+  })
```

Comment:
Should we also print the warnings in a batch instead of eagerly at the time of occurrence? They could easily get lost in the other warnings as the functions are getting executed.

I imagine a setup where at the time of printing the successful scores dt, we also print "In addition, the following metrics failed with the following warning". This is nicely discussed [in this blog post about deferring errors](https://reside-ic.github.io/blog/deferring-errors-with-r/). Worth considering.

Reply:
@jamesmbaazam we previously had something where errors would be saved. I don't have a clear picture for what this would look like exactly, but seems worth discussing. Would you be willing to create a new issue?